File: preppi_slim/SCR/run_method.pl
#!/usr/bin/perl

use strict;
use warnings;
use sigtrap 'handler', \&sig_handler, 'normal-signals';
use Getopt::Std;

use MODS::Methods;

if(!@ARGV) {
  print STDERR "Usage:  run_method.pl pipeline_name genome_name step_name gname sge_task_id\n";
  exit(-1);
  }

my $pname=shift;
my $gname=shift;
my $sname=shift;
my $tgtlst=shift;
my $taskid=shift;
my $quiet=shift;

my $pipeline=new MODS::Pipeline(name=>$pname,gname=>$gname);
my $genome=new MODS::Genome(gname=>$gname);
$tgtlst=`readlink -f $tgtlst`;
chomp $tgtlst;
my $jobdata=`perl -ne '(\$gid,\$nj)=split; if ($taskid<=\$alitot+\$nj) { chop; \$tid=$taskid-\$alitot; print "\$gid\t\$tid"; last; } \$alitot+=(split)[1];' $tgtlst`;
chomp $jobdata;
my ($gid,$sge_idx)=split(/\t/,$jobdata);
no strict 'refs';
my $mname="MODS::$sname";
my $method=$mname->new(gname=>$gname,gid=>$gid,step_parameters=>$pipeline->step_arg($sname));
$method->{sge_task_id}=$sge_idx if defined $method->{sge_input};
$method->{quiet}='no' if defined $quiet;
my $adir="$tgtlst.active";
$method->{active}="$adir/RUN.$gid.$sge_idx";
`touch $method->{active}; chmod g+rw $method->{active}` if -d "$tgtlst.active";
`touch $adir/LAST; chmod g+rw $adir/LAST` if defined $ENV{SGE_TASK_LAST} and $taskid==$ENV{SGE_TASK_LAST} and -d "$tgtlst.active";
my $tm=$pipeline->time();
chdir $method->{wrkdir} or die "Could not change directory to $method->{wrkdir}";
my $start=time();
$method->prep() if defined $quiet;
$method->run();
my $elapsed=time()-$start;
`touch $method->{wrkdir}/done; chmod g+rw $method->{wrkdir}/done`;
exit if not -d "$tgtlst.active";
`touch $adir/FREE.$gid.$sge_idx; chmod g+rw $adir/FREE.$gid.$sge_idx`;
my $ntasks=$method->count_tasks();
open(TIMFH,">$adir/TM.$gid.$sge_idx");
print TIMFH "$gid\t$sge_idx\t$elapsed\t$ntasks\n";
close TIMFH;
`chmod g+rw $adir/TM.$gid.$sge_idx`;
`echo "$tm" > $method->{wrkdir}/done; chmod g+rw $method->{wrkdir}/done`;
$tm=$pipeline->time();

File: preppi_slim/SCR/run_method.py
#!/usr/bin/env python3
import argparse
import os
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from MODS.Pipeline import Pipeline


def task_record(path, task_id):
    cumulative = 0
    for line in Path(path).read_text().splitlines():
        if not line.strip():
            continue
        target, count_text = line.split()
        count = int(count_text)
        if task_id <= cumulative + count:
            return target, task_id - cumulative
        cumulative += count
    raise ValueError(f"Task {task_id} is outside target list {path}")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("pipeline")
    parser.add_argument("genome")
    parser.add_argument("step")
    parser.add_argument("targets")
    parser.add_argument("task_id", type=int)
    parser.add_argument("debug", nargs="?")
    args = parser.parse_args()
    pipeline = Pipeline(name=args.pipeline, gname=args.genome)
    target_file = Path(args.targets).resolve()
    target, method_task = task_record(target_file, args.task_id)
    method = pipeline.get_step_class(args.step)(
        gname=args.genome, gid=target, step_parameters=pipeline.step_arg(args.step)
    )
    if hasattr(method, "sge_input"):
        method.sge_task_id = method_task
    if args.debug:
        method.quiet = "no"
        method.debug = "yes"
    active = Path(str(target_file) + ".active")
    run_marker = active / f"RUN.{target}.{method_task}"
    if active.is_dir():
        run_marker.touch()
    os.chdir(method.wrkdir)
    outputs = pipeline.method_outputs(method)
    pipeline.status_table.start(target, args.step, outputs)
    started = time.perf_counter()
    try:
        method.run()
    except BaseException as error:
        pipeline.status_table.fail(
            target,
            args.step,
            elapsed_seconds=time.perf_counter() - started,
            expected_outputs=outputs,
            error=f"{type(error).__name__}: {error}",
        )
        raise
    elapsed = time.perf_counter() - started
    Path(method.wrkdir, "done").write_text(pipeline.time() + "\n")
    if active.is_dir():
        (active / f"FREE.{target}.{method_task}").touch()
        (active / f"TM.{target}.{method_task}").write_text(
            f"{target}\t{method_task}\t{elapsed}\t{method.count_tasks()}\n"
        )
    pipeline.status_table.ready_for_postprocess(
        target,
        args.step,
        elapsed_seconds=elapsed,
        expected_outputs=outputs,
        message="Worker finished; output is awaiting verification and cleanup.",
    )


if __name__ == "__main__":
    main()
